### PR TITLE
Consistent domain count when domain creation fails

### PR DESCRIPTION
**Pull request by NickBarnes**
If domain creation fails, `caml_num_domains_running` still gets incremented, which means that `caml_domain_alone()` could (eventually) give the wrong answer. This PR moves the update of `caml_num_domains_running` to a point in `domain_create()` after the last failure case. It also does the same for the updates to the new domain's "interruptor" structure (which could in principle also be visible from outside the new domain, and which were not previously unwound in the failure paths).

Noticed while reviewing domain-creation code for #12855.

### CODE REVIEW

File: runtime/domain.c [hunk 1]
```diff
@@ -632,8 +628,7 @@ static void domain_create(uintnat initial_minor_heap_wsize,
   }
 
   domain_state->id = d->id;
```
Comment (ghost) [inline runtime/domain.c:630]: Shouldn't that assignment also get moved down to remain close to the assignment of `unique_id`?

Comment (NickBarnes) [reply runtime/domain.c:630]: Possibly, but I was uncertain whether any of the intervening function calls, into various modules, might use `domain_state->id` (for instance, in logging), whereas `unique_id` is not used much in the runtime and it was easy to determine that it was safe to move.

Comment (ghost) [reply runtime/domain.c:630]: Ok, let's keep `id` assigment location unchanged for now.

**Review by ghost — APPROVED** [review]

**Comment by NickBarnes** [comment]
I'll add a `Changes` entry if anyone thinks it merits one.

**Review by gasche — APPROVED** [review]
I agree that the code looks correct. We must really do this before registering the domain as a STW handler, but this is still the case after this PR.

**Comment by jmid** [comment]
Since this one goes hand in hand with #12855 I think it should also go into `5.2`.

**Comment by gasche** [comment]
Done.

**Pull request merged**